docs/spec/auth/subkeys_proposal.md

```diff
@@ -0,0 +1,156 @@
+
+One complexity in this system is that all txs, regardless of their msg type, need to be able to send tx fees. This means that even your lowest security subkey, if compromised, could drain your account by using your entire balance as the tx fee for a transaction. To resolve this, we add a notion of FeeAllowances.  The master pubkey, along with provisioning subkeys with msg types they are allowed to use, also provisions them with a daily allowance of tokens that they are allowed to spend to pay transaction fees.  For example, if SubKey A has a daily fee allowance of 1atom, it can spend up to 1atom on transaction fees in a 24 hour window.  If it does not use its atom in a 24 hour window, the "allowance" does not accumulate, it still stays at 1atom in any 24 hour window.
+
+
+## Construction
+
+To begin, we define the struct that stores the metadata about a certain account.  This includes the pubkey itself, the routes it's permitted to sign transactions for, its daily allowance, and how much of its allowance has been used in the last 24 hour window.
+
+```golang
+type SubKeyMetadata struct {
+  PubKey               sdk.AccPubKey
+  PermissionedRoutes   []string
+  DailyFeeAllowance    sdk.Coins
+  DailyFeeUsed         sdk.Coins
+}
+```
+
+We create a new `Account` type called `SubKeyAccount` that stores a list of `SubKeyMetadata`s.
+
+```golang
+type SubKeyAccount struct {
+  Address       AccAddress
+  Coins         Coins
+  PubKey        sdk.AccPubKey
+  AccountNumber uint64
+  Sequence      uint64
+  SubKeys       []SubKeyMetadata
+}
+```
+
+In the `StdSignature` type we need to add a new field called `PubKeyIndex` which tells the anteHandler which SubKey to use to verify this signature.
+
+```golang
+// StdSignature represents a sig
+type StdSignature struct {
+	crypto.PubKey                  `json:"pub_key"` // optional
+  Signature     []byte           `json:"signature"`
+  PubKeyIndex   uint
+}
+```
+
+If a `StdSignature.PubKeyIndex` is 0, this means that the account's "master" PubKey should be used to verify the signature.  If the `StdSignature.PubKeyIndex` is `> 0`, it verifies the signature using the corresponding **1-indexed** SubKey in the SubKeyAccount.SubKeys slice.  Note, the SubKeys slice is 1-indexed rather than 0-indexed because the 0 index is used to refer to the master PubKey.  See pseudocode below:
+
+```
+if stdSig.PubKeyIndex == 0 {
+  verify stdSig using acc.PubKey
+} else {
+  verify stdSig using acc.SubKeys[stdSig.PubKeyIndex - 1]
+}
+```
+
+In the AnteHandler, if a transaction comes from a SubKey, it verifies the msg route is permitted.
+
+```
+if msg.Route not in subKeyMetadata.PermissionedRoutes {
```

Comment:
There is no notion of messages in the ante handler. It only performs tx-level basic checks.

Reply:
Yeah, we'd need to have the AnteHandler inspect the msgs.  I think that means we have to pass the baseapp amino codec (where the msgs are declared) to the ante handler (if it's not already).